src/tools/compiletest/src/runtest.rs

```diff
@@ -1905,13 +1905,14 @@ impl<'test> TestCx<'test> {
         match output_file {
             TargetLocation::ThisFile(path) => {
                 rustc.arg("-o").arg(path);
+                rustc.arg("--out-dir").arg(&self.config.build_base);
```

Comment:
This is failing the UI test:
```
The actual stderr differed from the expected stderr.
Actual stderr saved to /home/jnelson/rust-lang/rust/build/x86_64-unknown-linux-gnu/test/ui/const-generics/const-argument-non-static-lifetime.full/const-argument-non-static-lifetime.full.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args const-generics/const-argument-non-static-lifetime.rs`

error in revision `full`: 1 errors occurred comparing output.
status: exit status: 0
command: "/home/jnelson/rust-lang/rust/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/home/jnelson/rust-lang/rust/src/test/ui/const-generics/const-argument-non-static-lifetime.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--cfg" "full" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-C" "prefer-dynamic" "-o" "/home/jnelson/rust-lang/rust/build/x86_64-unknown-linux-gnu/test/ui/const-generics/const-argument-non-static-lifetime.full/a" "--out-dir" "/home/jnelson/rust-lang/rust/build/x86_64-unknown-linux-gnu/test/ui/const-generics/const-argument-non-static-lifetime.full" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/home/jnelson/rust-lang/rust/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Zsave-analysis" "-L" "/home/jnelson/rust-lang/rust/build/x86_64-unknown-linux-gnu/test/ui/const-generics/const-argument-non-static-lifetime.full/auxiliary"
stdout: none
--- stderr -------------------------------
warning: ignoring --out-dir flag due to -o flag

warning: 1 warning emitted
------------------------------------------



failures:
    [ui] src/test/ui/const-generics/const-argument-non-static-lifetime.rs#full
```

I think the root issue is that `--out-dir` is *not* ignored by save-analysis, even if `-o` is also passed. I wonder if we should be using it at all - I see that there's some code in save-analysis that reads the config from an environment variable: https://github.com/rust-lang/rust/blob/b17e9d76f2ad15022e0e69bc33745c4ef9025a8f/compiler/rustc_save_analysis/src/lib.rs#L1006
I wonder if we should be using that instead, to avoid touching anything other than save-analysis itself? Try doing that instead of adding the `--out-dir` path and see if it works.

Reply:
@jyn514 
From the stderr, `--out-dir` is ignored:

```
warning: ignoring --out-dir flag due to -o flag
```

Reply:
@chenyukang the stderr is wrong. out-dir is ignored by most of the compiler when `-o` is passed, but not by save-analysis.

Reply:
@jyn514 Running the same rustc command as you with `--out-dir /home/ca/dev/rust/build/x86_64-unknown-linux-gnu/test/ui/const-generics/const-argument-non-static-lifetime.full` produces the save-analysis file `/home/ca/dev/rust/build/x86_64-unknown-linux-gnu/test/ui/const-generics/const-argument-non-static-lifetime.full/save-analysis/const_argument_non_static_lifetime.json` without the `save-analysis-temp` in the root directory as desired.

The test fails because of the stderr output.
Is this a problem?

Reply:
@jyn514 Changing `/home/ca/dev/rust/src/test/ui/const-generics/const-argument-non-static-lifetime.rs` to run-pass and running `x test src/test/ui/const-generics/const-argument-non-static-lifetime.rs` with `--out-dir /home/ca/dev/rust/build/x86_64-unknown-linux-gnu/test/ui/const-generics/const-argument-non-static-lifetime.full` as argument to rustc produces the save-analysis file `/home/ca/dev/rust/build/x86_64-unknown-linux-gnu/test/ui/const-generics/const-argument-non-static-lifetime.full/save-analysis/const_argument_non_static_lifetime.json` without `save-analysis-temp` in the root directory as desired.

The test fails because of the stderr output.
Is this a problem?

Reply:
@czzrr please ask these questions on Zulip or the tracking issue. My github notifications are a mess and I'm very unlikely to see any comment on a closed PR.